src/NServiceBus.Persistence.CosmosDB/TransactionalBatchDecoratorExtensions.cs

```diff
@@ -2,43 +2,58 @@
 
     static class TransactionalBatchDecoratorExtensions
     {
+        internal static async Task Execute(this TransactionalBatchDecorator transactionalBatch, Operation operation)
+        {
+            operation.Apply(transactionalBatch);
+
+            using (var batchOutcomeResponse = await transactionalBatch.Inner.ExecuteAsync().ConfigureAwait(false))
+            {
+                if (batchOutcomeResponse.Count > 1)
+                {
+                    throw new Exception("The transactional batch can only have one operation.");
+                }
+
+                var result = batchOutcomeResponse[0];
+
+                if (result.IsSuccessStatusCode)
+                {
+                    operation.Success(result);
+                    return;
+                }
+
+                // guaranteed to throw
+                operation.Conflict(result);
+            }
+        }
+
         internal static async Task Execute(this TransactionalBatchDecorator transactionalBatch, Dictionary<int, Operation> operationMappings)
         {
+            foreach (var operation in operationMappings.Values)
+            {
+                operation.Apply(transactionalBatch);
+            }
+
             using (var batchOutcomeResponse = await transactionalBatch.Inner.ExecuteAsync().ConfigureAwait(false))
             {
                 for (var i = 0; i < batchOutcomeResponse.Count; i++)
                 {
                     var result = batchOutcomeResponse[i];
 
-                    if (operationMappings.TryGetValue(i, out var operation))
-                    {
-                        if (result.IsSuccessStatusCode)
-                        {
-                            operation.Success(result);
-                            continue;
-                        }
-
-                        if (result.StatusCode == HttpStatusCode.Conflict || result.StatusCode == HttpStatusCode.PreconditionFailed)
-                        {
-                            // guaranteed to throw
-                            operation.Conflict(result);
-                        }
-                    }
+                    operationMappings.TryGetValue(i, out var operation);
+                    operation = operation ?? ThrowOnConflictOperation.Instance;
 
-                    if (result.StatusCode == HttpStatusCode.Conflict || result.StatusCode == HttpStatusCode.PreconditionFailed)
+                    if (result.IsSuccessStatusCode)
                     {
-                        throw new Exception("Concurrency conflict.");
+                        operation.Success(result);
+                        continue;
                     }
 
-                    if (result.StatusCode == HttpStatusCode.BadRequest)
-                    {
-                        throw new Exception("Bad request. Quite likely the partition key did not match");
-                    }
+                    // guaranteed to throw
+                    operation.Conflict(result);
```

Comment:
We throw on the first operation that failed. Does `TransactionalBatch` do the same, or should we somehow aggregate all of the results and throw that?

Reply:
I preserved the behaviour we already had in place. Happy to do what you suggest. Another PR?

Reply:
Sure.

Reply:
We can aggregate status codes of all the operations, but only the failing operation will have the error message. Every operation that did not execute as a result of the faulted operations will only have its status code set to [`424 - FailedDependency`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.cosmos.transactionalbatch.executeasync?view=azure-dotnet#Microsoft_Azure_Cosmos_TransactionalBatch_ExecuteAsync_System_Threading_CancellationToken_).

> If an operation within the transactional batch fails during execution, no changes from the batch will be committed and the status of the failing operation is made available in the StatusCode. To get more details about the operation that failed, the response can be enumerated - this returns TransactionalBatchOperationResult instances corresponding to each operation in the transactional batch in the order they were added into the transactional batch. For a result corresponding to an operation within the transactional batch, the StatusCode indicates the status of the operation - if the operation was not executed or it was aborted due to the failure of another operation within the transactional batch, the value of this field will be HTTP 424 (Failed Dependency); for the operation that caused the batch to abort, the value of this field will indicate the cause of failure as a HTTP status code.

Given that, I see no point in aggregation.